### PR TITLE
feat(update.jenkins.io): terraform generation of the infra around update.jenkins.io

### PR DESCRIPTION
**Pull request by smerle33**
In order to migrate update.jenkins.io (as per https://github.com/jenkins-infra/helpdesk/issues/2649), we need to provide as code the infrastructure to host it :
 - lan (public and internal subnets)
 - public ip
 - ubuntu 22.04 VM (4 vcpu, 16Gb)
 - datastore 1200Gb
 - firewall rules

### CODE REVIEW

File: Jenkinsfile_k8s [hunk 1]
```diff
@@ -6,20 +6,24 @@ parallel(
     // as per https://github.com/jenkins-infra/shared-tools/tree/main/terraform#jenkins-pipeline
     terraform(
       stagingCredentials: [
-        string(variable: 'ORACLE_OCI_CLI_USER', credentialsId: 'staging_oracle_oci_cli_user'),
-        string(variable: 'ORACLE_OCI_CLI_FINGERPRINT', credentialsId:'staging_oracle_oci_cli_fingerprint'),
+        string(variable: 'TF_VAR_region', credentialsId: 'oracle_oci_cli_region'),
+        string(variable: 'TF_VAR_tenancy_ocid', credentialsId: 'oracle_oci_cli_tenancy'),
```
Comment (dduportal) [inline Jenkinsfile_k8s:10]: Don't forget to add this secret in infra.ci ;)

Comment (smerle33) [reply Jenkinsfile_k8s:10]: already in with another name

File: updates.jenkins.io.tf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    memory_in_gbs = 16
+  }
+  source_details {
+    source_type = "image"
+    source_id   = data.oci_core_images.updates_jenkins_io.images[0].id
+  }
+  create_vnic_details {
+    subnet_id        = oci_core_subnet.public_subnet.id
+    assign_public_ip = true
+  }
```
Comment (dduportal) [inline updates.jenkins.io.tf:60]: todo: add private VNIC interface

File: updates.jenkins.io.tf [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  metadata = {
+    ssh_authorized_keys = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGFrPRIlP8qplANgNa3IO5c1gh0ZqNNj17RZeYcm+Jcb jenkins-infra-team@googlegroups.com"
```
Comment (dduportal) [inline updates.jenkins.io.tf:62]: todo:
- Damien to share the key in a GPG encrypted location with the team
- Add the network security group rules to allow incoming SSH to this machine
  - Step 1: from public IP to test the machine (restrict to our personal IPs as source) 
  - Step 2 (idea): use the private interface (VPN? SSH bastion) but later

File: updates.jenkins.io.tf [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  }
+  display_name  = "VM for updates.jenkins.io service"
+  freeform_tags = local.all_tags
+}
+
+resource "oci_core_volume_attachment" "updates_jenkins_io_data" {
+  # Paravirtualized attachment is expected to automount the data volume (compared to "iscsi")
+  attachment_type = "paravirtualized"
```
Comment (dduportal) [inline updates.jenkins.io.tf:81]: todo: 
- check if the volume is *really* automounted (case of Oracle documentation lying to us), and if yes: where is it mounted? Can we fine-tune the mount path here (to map to the actual production)

Comment (smerle33) [reply updates.jenkins.io.tf:81]: it is not mounted ... as for the path to mount, on the actual machine the "big" volume is on / ... which on this VM is reserved to the boot volume ...
the mount part, seems to need a partition step and I don't terraform option for that, I feel that we will need a go on cloud-init ....

Comment (smerle33) [reply updates.jenkins.io.tf:81]: as talked together, I just added a name for the volume : `/dev/oracleoci/oraclevdb`
and we will deal with the mount with puppet.

File: networks.tf [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  cidr_block                 = "10.0.0.0/24"
+  compartment_id             = var.compartment_ocid
+  vcn_id                     = oci_core_vcn.default.id
+  freeform_tags              = local.all_tags
+  prohibit_public_ip_on_vnic = false
+
+}
+
+# Sub network use to communicate machine to machine internally
+resource "oci_core_subnet" "internal_subnet" {
```
Comment (dduportal) [inline networks.tf]:
```suggestion
resource "oci_core_subnet" "private_subnet" {
```

nit: would oppose to "public", while internal can have different meanings (is it reachable? Is it internal to the VCN, to the datacenter, to the VM ?)

File: networks.tf [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  freeform_tags  = local.all_tags
+  private_ip_id  = data.oci_core_private_ips.updates_jenkins_io.private_ips[0].id
+}
+
+resource "oci_core_network_security_group" "updates_jenkins_io" {
+  compartment_id = var.compartment_ocid
+  vcn_id         = oci_core_vcn.default.id
+  display_name   = "updates.jenkins.io security group"
+  freeform_tags  = local.all_tags
+}
```
Comment (dduportal) [inline networks.tf]: Can you move this 3 definitions to the updates.jenkins.io.tf as it is only related to this service (and not to the general network settings) ?

File: networks.tf [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  route_rules {
+    network_entity_id = oci_core_internet_gateway.network_gateway.id
+    destination       = "0.0.0.0/0"
+    destination_type  = "CIDR_BLOCK"
+  }
+}
+
+output "instance_public_ip" {
+  value = oci_core_public_ip.updates_jenkins_io.ip_address
+}
```
Comment (dduportal) [inline networks.tf]:
```suggestion
output "updates_jenkins_io_public_ip" {
  value = oci_core_public_ip.updates_jenkins_io.ip_address
}
```
=> and also moving this definition into the file `update.jenkins.io.tf`

File: updates.jenkins.io.tf [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  }
+  create_vnic_details {
+    subnet_id        = oci_core_subnet.public_subnet.id
+    assign_public_ip = false #will assign a non ephemeral one (RESERVED ip)
+    nsg_ids          = [oci_core_network_security_group.updates_jenkins_io.id]
+  }
+  metadata = {
+    ssh_authorized_keys = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGFrPRIlP8qplANgNa3IO5c1gh0ZqNNj17RZeYcm+Jcb jenkins-infra-team@googlegroups.com"
+  }
+  display_name  = "VM for updates.jenkins.io service"
```
Comment (dduportal) [inline updates.jenkins.io.tf]:
```suggestion
  display_name  = "Virtual Machine for updates.jenkins.io service"
```

nit

**Review by dduportal — DISMISSED** [review]
A few feedbacks:

- Name feedback are all nitpicking
- Would be nice to move the definitions only related to updates.jenkins.io into the file `updates.jenkins.io.tf`


Otherwise great work, it sounds really good!

File: networks.tf (hunk 5 above)
Comment (smerle33) [reply networks.tf]: you did choose that name ... I though that you had a good reason 😂

**Review by dduportal — APPROVED** [review]
Plan of the last successfully build (build 41) says `No changes. Your infrastructure matches the configuration.`

=> you'll have to destroy resources and trigger again a new run on this PR + on main branch

- The build of the main branch must be successfull 
- This build of this PR must be successfull with a terraform plan showing all the resource that *would* be created.

**Review by dduportal — APPROVED** [review]
🤠 🚢

File: updates.jenkins.io.tf (hunk 3 above)
Comment (lemeurherve) [reply updates.jenkins.io.tf:62]: I might have missed it but where does the key has been stored in the end?

Comment (smerle33) [reply updates.jenkins.io.tf:62]: shared within slack in a gpg signed form

Comment (dduportal) [reply updates.jenkins.io.tf:62]: ideally, this key should only be used at bootstrap if really required, until the first puppet agent run is executed. Then we could (should!) delete it and forget it.

**Pull request merged**